etcd/tasks.go

```diff
@@ -2,26 +2,44 @@ package etcd
 
 import (
 	"github.com/giantswarm/microerror"
+	"time"
 )
 
 func FullBackup(b BackupInterface) error {
 	var err error
 
 	version := b.Version()
 
+	start := time.Now()
+
 	err = b.Create()
 	if err != nil {
 		return microerror.Maskf(err, "Etcd %s creation failed: %s", version, err)
 	}
 
+	creationTime := time.Since(start).Nanoseconds() / 1e6
+
+	start = time.Now()
+
 	err = b.Encrypt()
 	if err != nil {
 		microerror.Maskf(err, "Etcd %s encryption failed: %s", version, err)
 	}
 
-	err = b.Upload()
+	encryptionTime := time.Since(start).Nanoseconds() / 1e6
+	start = time.Now()
+
+	size, err := b.Upload()
 	if err != nil {
 		microerror.Maskf(err, "Etcd %s upload failed: %s", version, err)
 	}
+
+	uploadTime := time.Since(start).Nanoseconds() / 1e6
+
+	err = b.SendMetrics(creationTime, encryptionTime, uploadTime, size)
+	if err != nil {
+		microerror.Maskf(err, "Etcd %s metrics push failed: %s", version, err)
+	}
+
```

Comment:
sending metrics here is tricky, in theory, it should be fine but its in the whole FullBackup function which is here: https://github.com/giantswarm/etcd-backup/blob/master/service/service.go#L215-L232 

run inside of backoff library which is doing a retry logic in case of failure and if for any unspecified reason the function fail at the end there might be a scenario where you try send the metrics two times and that might cause error as proetheus dont allow duplicate metrics with same values

It might help to move send Metrics call at the end of the retry function here: https://github.com/giantswarm/etcd-backup/blob/master/service/service.go#L218

if there is and `err==nil` then you would collect metrics for success backup and if `err!=nil` then you can collect metrics for failed backup

Reply:
I'm not following you here. If I move the metrics sending code in that point, in case of a failure the `o` function will be retried by the `backoff` thing and the metrics would be pushed more than once (as it happens now). To reach the desired result, I could send the success metrics only from inside the `o` function and send the failure metrics only if the whole `backoff` system fails. What do you think?